### PR TITLE
ATen scheduler for the new `Matmul/LinearOp` IR nodes

### PR DESCRIPTION
**Pull request by Priya2698**
1. Adds a new scheduler -- `ExprEvalScheduler` that accepts the MatmulOp and LinearOp (next PR) for ATen evaluation.
2. Modify the matmul input generator to test for all cases supported by Thunder.
3. The `eagerMatmul` API is renamed and replaces the existing `matmul` API. `fd.ops.matmul` now creates a `MatmulOp` (except in a few special cases such as scalar dot product, for eg: `[K] x [K]`.

Issue #2149, #2092.

### CODE REVIEW

File: csrc/scheduler/expr_eval_sched.h
Comment (jacobhinkle) [inline csrc/scheduler/expr_eval_sched.h:50]: I'm not sure what the comment here means.

File: csrc/scheduler/expr_eval_sched.h
Comment (jacobhinkle) [inline csrc/scheduler/expr_eval_sched.h:46]: I don't think this is needed.

File: csrc/scheduler/expr_eval_sched.h
Comment (jacobhinkle) [reply csrc/scheduler/expr_eval_sched.h:50]: Can you just use `HeuristicParams` in place of `ExprEvalHeuristic` or is the problem that that class is pure virtual. You _could_ consider just providing trivial defaults for those virtual methods then you wouldn't need this class.

File: csrc/scheduler/expr_eval_sched.h
Comment (jacobhinkle) [inline csrc/scheduler/expr_eval_sched.h:61]: Could do this to avoid creating a new `shared_ptr`
```suggestion
    auto* other_casted = dynamic_cast<ReductionParams*>(other.get());
```
Also, `HeuristicParams` is a `PolymorphicBase` so you can use
```c++
  if (!other->isA<ReductionParams>()) {
    return false;
  }
  return other->cparams == cparams;
}
```

**Review by jacobhinkle — COMMENTED** [review]
Looking great. Sprinkle in a few tests once you merge #2175 and we'll be on our way.

**Comment by Priya2698** [comment]
!build

File: csrc/ops/composite.h
Comment (jacobhinkle) [inline csrc/ops/composite.h:74]: I suggest keeping the documentation in place.
```suggestion
// Matmul function which takes in tensors with the shapes
// A[M,K] B[K,N], but the tensors may have different layouts
// via strides. All restrictions from the matmul APIs also
// apply here.
TensorView* matmul(TensorView* tv_a, TensorView* tv_b);
```

File: csrc/root_domain_map.cpp
Comment (jacobhinkle) [inline csrc/root_domain_map.cpp:188]: Good change. These names are much clearer.

File: csrc/root_domain_map.cpp
Comment (jacobhinkle) [inline csrc/root_domain_map.cpp:175]: Why are these being removed? Doesn't this cause breakage?

File: csrc/scheduler/expr_eval_sched.cpp
Comment (jacobhinkle) [inline csrc/scheduler/expr_eval_sched.cpp:40]: nit: I would place this in the header.

File: csrc/scheduler/expr_eval_sched.cpp
Comment (jacobhinkle) [inline csrc/scheduler/expr_eval_sched.cpp:21]: Do you need this line? Also: I would place this in the header.

File: csrc/scheduler/heuristic.h
Comment (jacobhinkle) [inline csrc/scheduler/heuristic.h:33]: You might want `other->isStrictlyA<HeuristicParams>()` to do a strict check. Otherwise this will be false unless other is `nullptr`.

File: csrc/scheduler/heuristic_types.h
Comment (jacobhinkle) [inline csrc/scheduler/heuristic_types.h:64]: Should `NoOp` come before `ExprEval`?

File: tests/python/pytest_input_generators.py
Comment (jacobhinkle) [inline tests/python/pytest_input_generators.py:1515]: Can we sprinkle in some broadcast dims?

File: version.txt
Comment (jacobhinkle) [inline version.txt:1]: :+1:

**Review by jacobhinkle — CHANGES_REQUESTED** [review]
Looks good overall. I am just slightly concerned about removing stuff from `PairwiseRootDomainMap::map`.

File: csrc/root_domain_map.cpp
Comment (Priya2698) [reply csrc/root_domain_map.cpp:175]: I moved them within `updatePairwiseDomainMap` since that will be a common check for any special cases. The other 3 conditions were op-specific (select, torch gather etc).

File: csrc/ops/composite.h
Comment (Priya2698) [reply csrc/ops/composite.h:74]: Added a comment detailing the functionality. The above comment is incorrect since we support all sizes now.

File: csrc/root_domain_map.cpp
Comment (jacobhinkle) [reply csrc/root_domain_map.cpp:175]: Ah ok of course. I missed that this is called in those cases now too.

File: csrc/scheduler/expr_eval_sched.cpp
Comment (jacobhinkle) [reply csrc/scheduler/expr_eval_sched.cpp:21]: Interesting. I wonder if you could just force `DataType::Int` for the index type always with this scheduler. Note I'm not exactly sure how to do that and it might be more work than this single line.

Comment (jacobhinkle) [reply csrc/scheduler/expr_eval_sched.cpp:21]: Nevermind, I think this line is fine. You _could_ ignore the runtime_info index type and force int64, but that's orthogonal to whether this line should exist. I think you're good to go.

**Review by jacobhinkle — APPROVED** [review]
LGTM after tests pass and you add some broadcasts in test.

File: csrc/scheduler/heuristic_types.h
Comment (Priya2698) [reply csrc/scheduler/heuristic_types.h:64]: Some cases get accepted by NoOp scheduler, which is why I prioritized ExprEval scheduler.

We may need to change the heuristics of `NoOp` if we want to switch the order.

**Comment by Priya2698** [comment]
!build

**Comment by jjsjann123** [comment]
:rocket:

File: csrc/scheduler/heuristic_types.h
Comment (jacobhinkle) [reply csrc/scheduler/heuristic_types.h:64]: Oh! Thanks for mentioning that. Does `NoOp` scheduler accept the cases where you have a single scalar output? Because it seems to me that it would do so based on this code: https://github.com/NVIDIA/Fuser/blob/8c18701e252f279292f9ae1c926b7e919b3ebd17/csrc/fusion.cpp#L341-L359
We should add a special case for zero-dimensional outputs there! I'll send a PR in a moment that does that and flips NoOp to be first again.

Comment (jacobhinkle) [reply csrc/scheduler/heuristic_types.h:64]: https://github.com/NVIDIA/Fuser/issues/2241

Comment (Priya2698) [reply csrc/scheduler/heuristic_types.h:64]:
```
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] ATenNodesParametrizedTest.MatmulNodeConcrete/2, where GetParam() = ({ 32 }, { 32, 1 })
[  FAILED  ] ATenNodesParametrizedTest.MatmulNodeConcrete/8, where GetParam() = ({ 1, 32 }, { 32 })
[  FAILED  ] ATenNodesParametrizedTest.MatmulNodeConcrete/10, where GetParam() = ({ 1, 32 }, { 32, 1 })
[  FAILED  ] ATenNodesParametrizedTest.MatmulNodeSymbolic/2, where GetParam() = ({ 32 }, { 32, 1 })
[  FAILED  ] ATenNodesParametrizedTest.MatmulNodeSymbolic/8, where GetParam() = ({ 1, 32 }, { 32 })
[  FAILED  ] ATenNodesParametrizedTest.MatmulNodeSymbolic/10, where GetParam() = ({ 1, 32 }, { 32, 1 })
```

It is likely because there no reductions identified since we use ATen, and all the dimensions in the output are broadcast dimensions. So the cases where M/N = 1 get picked by `NoOp`

**Pull request merged**
